app/presenters/post_presenter.rb

```diff
@@ -3,82 +3,93 @@ class PostPresenter
+      open_graph_cache:             build_open_graph_cache,
+      mentioned_people:             build_mentioned_people_json,
+      photos:                       build_photos_json,
+      root:                         root,
+      title:                        title,
+      address:                      @post.address,
+      poll:                         @post.poll,
+      already_participated_in_poll: already_participated_in_poll,
+      participation:                participate?,
+      interactions:                 build_interactions_json
+    }
+  end
+
+  def build_text
+    if @post.message
       @post.message.plain_text_for_json
     else
       @post.raw_message
     end
-    {
-        :id => @post.id,
-        :guid => @post.guid,
-        :text => text,
-        :public => @post.public,
-        :created_at => @post.created_at,
-        :interacted_at => @post.interacted_at,
-        :provider_display_name => @post.provider_display_name,
-        :post_type => @post.post_type,
-        :image_url => @post.image_url,
-        :object_url => @post.object_url,
-        :nsfw => @post.nsfw,
-        :author => @post.author.as_api_response(:backbone),
-        :o_embed_cache => @post.o_embed_cache.try(:as_api_response, :backbone),
-        :open_graph_cache => @post.open_graph_cache.try(:as_api_response, :backbone),
-        :mentioned_people => @post.mentioned_people.as_api_response(:backbone),
-        :photos => @post.photos.map {|p| p.as_api_response(:backbone)},
-        :root => root,
-        :title => title,
-        :address => @post.address,
-        :poll => @post.poll(),
-        :already_participated_in_poll => already_participated_in_poll,
-        :participation => participate?,
-
-        :interactions => {
-            :likes => [user_like].compact,
-            :reshares => [user_reshare].compact,
-            :comments_count => @post.comments_count,
-            :likes_count => @post.likes_count,
-            :reshares_count => @post.reshares_count
-        }
-    }
   end
 
-  def title
-    @post.message.present? ? @post.message.title : I18n.t('posts.presenter.title', name: @post.author_name)
+  def build_open_graph_cache
+    @post.open_graph_cache.try(:as_api_response, :backbone)
   end
 
-  def root
-    PostPresenter.new(@post.absolute_root, current_user).as_json if @post.respond_to?(:absolute_root) && @post.absolute_root.present?
+  def build_mentioned_people_json
+    @post.mentioned_people.as_api_response(:backbone)
   end
 
-  def user_like
-    @post.like_for(@current_user).try(:as_api_response, :backbone)
+  def build_photos_json
+    @post.photos.map {|p| p.as_api_response(:backbone) }
   end
 
-  def user_reshare
-    @post.reshare_for(@current_user)
+  def title
+    @post.message.present? ? @post.message.title : I18n.t("posts.presenter.title", name: @post.author_name)
   end
 
-  protected
+  def root
+    if @post.respond_to?(:absolute_root) && @post.absolute_root.present?
```

Comment:
Should I make this into `return unless (@post.respond_to?(:absolute_root) || @post.absolute_root.present?)` ? It would be harder to read for me.